### DISCUSSION TITLE
Any IR read/write head with the meter "EasyMeter Q3DB1024 V3.04" in operation?

### DISCUSSION THREAD

**Discussion by fkroepfl** [arendst/Tasmota#21405] [General]
Does anyone have experience with IR read/write heads and the "EasyMeter Q3DB1024 V3.04" meter?

If so, please share which IR read/write head or which, possibly customized, script is used?

Many thanks in advance.

**Comment by HofeBY** [comment]
https://tasmota.github.io/docs/Smart-Meter-Interface/#easymeter-q3d-q3da1024-obis

**Reply by fkroepfl** [reply]
Thank you very much!
Do you have a setup with the unchanged script in operation? 
Which IR head have you connected or are you using for reading the meter data?

**Reply by HofeBY** [reply]
1. No, not fully ready.

2.
https://www.ebay.de/itm/285350331996

https://de.aliexpress.com/item/1005006851575663.html?spm=a2g0o.productlist.main.115.351299540YEkf0&algo_pvid=9429d01b-5049-409b-b371-8ea9cb6b615f&algo_exp_id=9429d01b-5049-409b-b371-8ea9cb6b615f-57&pdp_npi=4%40dis%21EUR%218.91%215.88%21%21%219.35%216.17%21%402103834817151611955598328ec45b%2112000038515387241%21sea%21DE%213001189247%21&curPageLogUid=4XldGZR1IKn5&utparam-url=scene%3Asearch%7Cquery_from%3A

or with nice WS2812 support

https://de.aliexpress.com/item/1005005569520224.html?spm=a2g0o.productlist.main.81.351299540YEkf0&algo_pvid=9429d01b-5049-409b-b371-8ea9cb6b615f&algo_exp_id=9429d01b-5049-409b-b371-8ea9cb6b615f-40&pdp_npi=4%40dis%21EUR%216.96%216.96%21%21%2152.76%2152.76%21%402103834817151611955598328ec45b%2112000034319347393%21sea%21DE%213001189247%21&curPageLogUid=4VoW7DqowrBS&utparam-url=scene%3Asearch%7Cquery_from%3A

In the hope of fully Tasmota support matter & thread & Wifi 6 & BT 5.3 (LE) & Zigbee on this Boardtype.

HofeBY

**Reply by fkroepfl** [reply]
May ask for an update when you are done with it and it works. 🤗

**Reply by HofeBY** [reply]
what language ?

Oberfränkische Grüße
HofeBY

**Reply by fkroepfl** [reply]
English is - I think - OK in terms of forum language or if it should be native, then we can also consider Austrian 😉 .

**Reply by HofeBY** [reply]
OMG

I think Austrian is not supported Language by Tasmota. ;-)

You have to accept german.

https://github.com/HofeBY/Tasmota-Firmware

HofeBY

**Reply by fkroepfl** [reply]
Thank you very much. 🤗

**Reply by fkroepfl** [reply]
@HofeBY 

Question for understanding: According to https://github.com/HofeBY/Tasmota-Firmware/blob/main/Liesmich-zuerst.txt, GPIO1 and GPIO3 are connected to the USB-Connecot.

Can these GPIOs then be used for the serial adapter, be it IR or Mbus, as a connection to the smart meter or do you have to switch to other GPIOs and if you have to switch, which GPIOs are recommended or should be used?

The background to the question is the comment at https://tasmota.github.io/docs/Smart-Meter-Interface/#easymeter-q3b-sml

**Reply by fkroepfl** [reply]
Additional question: Is it possible to read all the other meters listed with your precompiled binary files, even if SML is not explicitly mentioned for these types?

Example: https://tasmota.github.io/docs/Smart-Meter-Interface/#2-landisgyr-e450-encrypted

**Reply by HofeBY** [reply]
GPIO1 & GPIO3 can be used elsewhere if it is no more connected to USB to TTY Chip for device programming.
But is this a good idea until you have enough free lines ?

GPIO1 & GPIO3 in my example config is used as serial connetion via USB Adaptor to Putty on my Laptop to see what´s
going on in LOG of ESPxxx & I can also send commands by USB to Tasmota. Do not think to much about that. Just additional.
I also config an I2C Bus to connect several Sensors & IO Multyplayer.
Also the Hall & CPUtemp  config is also just to show the possibility´s and to use it in examples.
Nothing to do with SML, but also usable.

All "CPU.bin" Variants including SML new Version 14.0.0.1 has been build.
I´m shure, that any version of Meter exist, that are not beeing supportet.
Try out and give response which is missing.
Description to build for your own Version are there.
read "liesmich-zuerst.txt" and follow it.
If you find out new things to include give me the solution, so I can include it too.

Maybe next Version include Script examples for Meter and sensors and also "MATTER" version´s.
SML has the need for SCRIPT and compatibility to ESP8285 & ESP8266 . But MATTER can not be used with SCRIPT on the same device.

HofeBY

**Reply by fkroepfl** [reply]
The reason for looking at GPIO1&3 was this comment ...

`Because the 7E1 serial mode is not supported by Tasmota software serial, the hardware serial port must be used, i.e. GPIO 3. This will /not/ work using GPIO 0 or 2.
`
On the other hand it seems that GPIO16&17 are as well hardware serial ports, or? 
 
![image](https://github.com/arendst/Tasmota/assets/6333880/7461a7b4-5b37-4ae4-91d2-fa3acb42eeb0)

**Reply by fkroepfl** [reply]
> https://github.com/HofeBY/Tasmota-Firmware

Apparently there is a limitation in the official Tasmota version that prevents the data from being read in this case.

- https://github.com/arendst/Tasmota/discussions/21359#discussioncomment-9510407

```
@@ -2007,8 +2024,10 @@ void SML_Decode(uint8_t index) {
               } else {
                 mp++;
                 if (isdigit(*mp)) {
-                  uint8_t skip = strtol((char*)mp, (char**)&mp, 10);
+                  uint16_t skip = strtol((char*)mp, (char**)&mp, 10);
                   cp += skip;

```

@HofeBY Could you provide me with a modified binary file for ESP32 and ESP32V3?
Unfortunately, I have no experience of how to do this. Sorry. 🤦‍♂️

**Reply by HofeBY** [reply]
You mean ESP32V3 (I don´t know this type) or ESP32C3 ?
I upload ESP32C3 firmware too.

**Reply by fkroepfl** [reply]
Sorry, ESP32S3 was meant.

**Reply by fkroepfl** [reply]
These are the 2 boards I have available for my tests ... 😉

- ESP Chip ID | 15053036 (ESP32-S3 v0.2)
- ESP Chip ID | 8474100 (ESP32-D0WDQ6-V3 v3.0)

**Comment by fkroepfl** [comment]
I have installed with "Upgrade by web server" and https://github.com/HofeBY/Tasmota-Firmware/blob/main/tasmota32.bin, at least I hope so, but it seems that https://tasmota.github.io/docs/Scripting-Language/ is not working as expected.

```
07:26:02.175 CMD: edit script
07:26:02.180 RSL: RESULT = {"Command":"Unknown"}
```

What could be the problem?

**Reply by HofeBY2** [reply]
Oh well.
EDIT(.exe) is an poor Windows Text Editor. Not existing in Tasmota.
After Installing tasmota32.bin from my Gitubsite and WLAN go install -Werkzeuge - Script

**Reply by fkroepfl** [reply]
With the direct upload of the BIN file it seems to have worked now ...

<img width="875" alt="Bildschirmfoto 2024-05-22 um 10 59 04" src="https://github.com/arendst/Tasmota/assets/6333880/dccf03db-e354-4faa-89e3-8339ee7d7192">

**Reply by HofeBY2** [reply]
Go on

**Reply by HofeBY2** [reply]
Have you find out with your Handycamera which of that to Diods of Meter & Hichshead are Transmitter and Receiver ?
Transmitter to Receiver !

**Reply by fkroepfl** [reply]
I will try to connect the existing IR read head, probably with RS232 interface, via level converter.
Probably because the manufacturer is unknown and some things are based on assumptions.
If this does not work, a new IR read head will have to be purchased.

**Reply by HofeBY2** [reply]
It's serial TTL 3,3V and not RS232 !
What about Handycam test ?

**Reply by fkroepfl** [reply]
Thanks for thinking along, but with the "level converter" I meant a RS232 MAX3232 TO TTL CONVERTER MODULE, which is connected between the external RS232 and the internal 3.3V TTL interface ports. IMHO that should work fine. At least as long my soldering is OK. 🤗

As I'm currently not on site, I was not able to run any local tests. 😉

**Reply by sfromis** [reply]
Are you sure that your thing is actually RS232 (+/- 12V) which is uncommon for those IR heads, instead of the usual TTL UART (levels 0 or 3.3V)?

**Reply by fkroepfl** [reply]
No, I'm not entirely sure. Unfortunately there is nothing, absolutely nothing, printed on the read head.

In a purely graphical "design comparison" on the Internet, I only came across similar designs here: 
 
- https://emh-metering.com/produkte/kommunikation/okk-bke-gen-f/ resp. 
 
- https://www.energie-zaehler.com/epages/61422236.mobile/de_DE/?ObjectPath=/Shops/61422236/Products/OPKEMRS232

The indication of RS232 comes from the 4-pin RJ10 (4P4C) western plug used, which is sometimes used for RS232.

**Reply by HofeBY** [reply]
Your reading Head is only on the backside of eHz Meter usefull, and has to be installed before Meterinstallation.
save time and money by ordering a better concept.

complete:
https://www.ebay.de/itm/126141861155

TTL with Case and Magnet:
https://www.ebay.de/itm/314015465828?_trkparms=amclksrc%3DITM%26aid%3D1110025%26algo%3DHOMESPLICE.COMPOSITELISTINGS%26ao%3D1%26asc%3D265705%26meid%3Dbf69af51fe3247e99381f628f0a6bfab%26pid%3D101506%26rk%3D9%26rkt%3D25%26sd%3D285350331996%26itm%3D314015465828%26pmt%3D1%26noa%3D1%26pg%3D4247815%26algv%3DAlgoIndex5SimRanker&_trksid=p4247815.c101506.m1851

TTL with Magnet:
https://www.ebay.de/itm/354322386156?_trkparms=amclksrc%3DITM%26aid%3D1110025%26algo%3DHOMESPLICE.COMPOSITELISTINGS%26ao%3D1%26asc%3D265705%26meid%3Dbf69af51fe3247e99381f628f0a6bfab%26pid%3D101506%26rk%3D10%26rkt%3D25%26sd%3D285350331996%26itm%3D354322386156%26pmt%3D1%26noa%3D1%26pg%3D4247815%26algv%3DAlgoIndex5SimRanker&_trksid=p4247815.c101506.m1851

TTL pure:
https://www.ebay.de/itm/285350331996

**Reply by fkroepfl** [reply]
> TTL with Case and Magnet:

Many thanks for the tip. I really like this variant. 🤗

**Comment by fkroepfl** [comment]
Many thanks to everyone, I have now found a working solution. => https://github.com/arendst/Tasmota/discussions/21576#discussion-6785702
